memcache/selector_test.go

```diff
@@ -16,24 +16,284 @@ limitations under the License.
-	var ss ServerList
-	ss.SetServers(servers...)
-	for i := 0; i < b.N; i++ {
-		if _, err := ss.PickServer("some key"); err != nil {
-			b.Fatal(err)
+func TestRetriesAfterError(t *testing.T) {
+	ss := createWithServers(t, []string{"127.0.0.1:1234"})
+
+	addrA := ss.addrs[0]
+	ss.OnResult(addrA, &net.OpError{})
+	_, err := ss.PickServer("hi")
+	assertEqualError(t, err, "memcache: no servers configured or available")
+
+	time.Sleep(startingWait * 2)
+	_, err = ss.PickServer("hi")
+	// should have become available again
+	requireNoError(t, err)
+}
+
+func TestRecoversOnSuccessfulCommunication(t *testing.T) {
+	ss := createWithServers(t, []string{"127.0.0.1:1234"})
+	addrA := ss.addrs[0]
+
+	ss.OnResult(addrA, &net.OpError{})
+	_, err := ss.PickServer("hi")
+	assertEqualError(t, err, "memcache: no servers configured or available")
+
+	ss.OnResult(addrA, nil)
+	_, err = ss.PickServer("hi")
+	// should have become available again
+	requireNoError(t, err)
+}
+
+func TestOneServer(t *testing.T) {
+	functionalTest(t, []string{"42111"})
+}
+
+func TestTwoServers(t *testing.T) {
+	functionalTest(t, []string{"42111", "42222"})
+}
+
+// This is a functional test that spins up and repeatedly takes down a memcached nodes
+// while multiple writer routines attempt to Set keys
+func functionalTest(t *testing.T, ports []string) {
+	srvs := []string{}
+	for _, port := range ports {
+		srvs = append(srvs, fmt.Sprintf("localhost:%s", port))
+	}
+	client := New(srvs...)
+
+	wg := sync.WaitGroup{}
+
+	server := func(port string) {
+		for i := 0; i < 5; i++ {
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second*3)
+			defer cancel()
+			cmd := exec.CommandContext(ctx, "memcached", "-p", port)
+			err := cmd.Run()
+			assertEqualError(t, err, "signal: killed")
+
+			// no randomisation - keep approximately in lockstep to keep the amount of downtime high
+			time.Sleep(time.Millisecond*500)
+		}
+		wg.Done()
+	}
+
+	done := make(chan struct{})
+	mx := sync.Mutex{}
+	resultCounts := map[string]int{}
+	networkErrorCategory := "[network error]"
+	internalTimeoutError := "[memcachepl timeout]"
+	writer := func() {
+		for {
+			select {
+			case <-done:
+				break
+			default:
+				err := client.Set(&Item{
+					Key:   "hi",
+					Value: []byte("val"),
+				})
+				errString := fmt.Sprintf("%v", err)
+				var netErr net.Error
+				if errors.As(err, &netErr) {
+					errString = networkErrorCategory
+				}
+				var cte *ConnectTimeoutError
+				if errors.As(err, &cte) {
+					errString = internalTimeoutError
+				}
+				mx.Lock()
+				resultCounts[errString]++
+				mx.Unlock()
+			}
+			time.Sleep(time.Duration(rand.Int31n(5)) * time.Millisecond)
+		}
+	}
+
+	// start servers
+	wg.Add(len(srvs))
+	for _, port := range ports {
+		go server(port)
+	}
+
+	go (func() {
+		wg.Wait()
+		close(done)
+	})()
+
+	// start writers
+	go writer()
+	go writer()
+	go writer()
+	go writer()
+	go writer()
+	go writer()
+
+	// wait for test run to end
+	wg.Wait()
```

Comment:
is there a reason why we can't just `close(done)` here? (not sure why we need the goroutine spawned on line 211).

Reply:
👍 good point